internal/resolution/manifest/maven.go

```diff
@@ -76,6 +91,56 @@ func (m MavenManifestIO) Read(df lockfile.DepFile) (Manifest, error) {
 		}
 	}
 
+	interpolate := func(project *maven.Project) error {
+		// Interpolating can change a dependency's key if it contained properties.
+		// If it is changed we need to update the requirementOrigins map with the new key.
+		allKeys := func() []maven.DependencyKey {
+			// Get all the dependencyKeys of the project.
+			// This order shouldn't change after calling Interpolate.
+			var keys []maven.DependencyKey
+			for _, dep := range project.Dependencies {
+				keys = append(keys, dep.Key())
+			}
+			for _, dep := range project.DependencyManagement.Dependencies {
+				keys = append(keys, dep.Key())
+			}
+			for _, profile := range project.Profiles {
+				for _, dep := range profile.Dependencies {
+					keys = append(keys, dep.Key())
+				}
+				for _, dep := range profile.DependencyManagement.Dependencies {
+					keys = append(keys, dep.Key())
+				}
+			}
+			for _, plugin := range project.Build.PluginManagement.Plugins {
+				for _, dep := range plugin.Dependencies {
+					keys = append(keys, dep.Key())
+				}
+			}
+
+			return keys
+		}
+
+		prevKeys := allKeys()
+		if err := project.Interpolate(); err != nil {
+			return err
+		}
+
+		newKeys := allKeys()
+		if len(prevKeys) != len(newKeys) {
+			return errors.New("number of dependencies changed after interpolation")
```

Comment:
the number of dependencies may get changed after interpolation if we are not able to resolve any placeholder, but this should be very rare...

Reply:
It's kind of hard to know which dependencies are which in those cases, so I'll leave it as an error for now.

Added a comment here.